src/core/mapping/base_mapper.cc

```diff
@@ -208,20 +209,35 @@ void BaseMapper::select_task_options(const Legion::Mapping::MapperContext ctx,
                                      const Legion::Task& task,
                                      TaskOptions& output)
 {
+  Task legate_task(&task, context, runtime, ctx);
 #ifdef LEGATE_USE_COLLECTIVE
-  for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
-    auto& req = task.regions[idx];
+  for (auto& store : legate_task.inputs()) {
+    if (store.is_future()) continue;
+    auto idx = store.requirement_index();
+    auto req = task.regions[idx];
+    if (req.privilege & LEGION_WRITE_PRIV) continue;
+    std::vector<int32_t> promoted_dims;
+    store.return_promoted_dims(promoted_dims);
+    for (auto& d : promoted_dims) {
+      if (d < 0) continue;
+      if ((task.index_domain.hi()[d] - task.index_domain.lo()[d]) >= 1) {
+        output.check_collective_regions.insert(idx);
+        break;
+      }
+    }
+  }
+  for (auto& store : legate_task.reductions()) {
+    if (store.is_future()) continue;
+    auto idx = store.requirement_index();
+    auto req = task.regions[idx];
     if (req.privilege & LEGION_WRITE_PRIV) continue;
-    // Look up the projection for the input region. There are cases where
-    // Legate libraries register their own projection functors that are
-    // not recorded by Legate Core. So, handle the case when these functors
-    // are not present and allow for them to be missing.
     auto projection = find_legate_projection_functor(req.projection, true /* allow_mising */);
     if ((req.handle_type == LEGION_SINGULAR_PROJECTION) ||
-        (projection != nullptr && projection->is_collective())) {
+        ((projection != nullptr) && (projection != 0))) {
```

Comment:
If all you need to do is check for non-identity projection functor, I think you just need to:

```
    if (req.handle_type == LEGION_SINGULAR_PROJECTION || req.projection != 0) {
```

And then presumably you don't need to do `find_legate_projection_functor` at all.

Notably, the `projection != 0` check that exists currently doesn't do anything, since you've already checked the equivalent  `projection != nullptr`.

Reply:
that is right, thank you